tests/bazel_tests.bzl

```diff
@@ -13,83 +13,118 @@ build:isolate --fetch=False
+    register = 'go_register_toolchains()\n'
+  elif ctx.attr.go_version != None:
+    register = 'go_register_toolchains(go_version="{}")\n'.format(ctx.attr.go_version)
+
+  workspace_content = 'workspace(name = "bazel_test")\n\n'
   for ext in ctx.attr.externals:
     root = ext.label.workspace_root
-    _,_,ws = root.rpartition("/")
-    workspace_content += 'local_repository(name = "{0}", path = "{1}/{2}")\n'.format(ws, ctx.attr._execroot.path, root)
-  # finalise the workspace file
-  workspace_content += 'load("@io_bazel_rules_go//go:def.bzl", "go_rules_dependencies", "go_register_toolchains")\n'
-  workspace_content += 'load("@io_bazel_rules_go//go/private:toolchain.bzl", "go_local_sdk")\n'
-  if ctx.attr.go_version == CURRENT_VERSION:
-    workspace_content += 'go_register_toolchains()\n'
-  elif ctx.attr.go_version:
-    workspace_content += 'go_register_toolchains(go_version="{}")\n'.format(ctx.attr.go_version)
+    _,_,name = ext.label.workspace_root.rpartition("/")
+    workspace_content += 'local_repository(name="{name}", path="{exec_root}/{root}")\n'.format(
+      name = name,
+      root = root,
+      exec_root = ctx.attr._settings.exec_root,
+    )
   if ctx.attr.workspace:
     workspace_content += ctx.attr.workspace
-  workspace_file = ctx.new_file(subdir + "WORKSPACE")
-  ctx.file_action(output=workspace_file, content=workspace_content)
-  # finalise the script
-  args += ctx.attr.args + ctx.attr.targets
-  script_content += 'BASE=$(pwd)\n'
-  script_content += 'cd {0}\n'.format(ctx.label.package)
-  script_content += 'PACKAGE=$(pwd)\n'
-  if ctx.attr.subdir:
-    script_content += 'cd {0}\n'.format(ctx.attr.subdir)
-    script_content += 'cp BUILD.in BUILD.bazel\n'
-  script_content += 'WORKSPACE=$(pwd)\n'
-  if ctx.attr.prepare:
-    script_content += ctx.attr.prepare
-  script_content += 'cd $WORKSPACE\n'
-  script_content += 'echo {0} {1}\n'.format(ctx.attr._execroot.bazel, " ".join(args))
-  script_content += '{0} {1}\n'.format(ctx.attr._execroot.bazel, " ".join(args))
-  script_content += 'result=$?\n'
-  if ctx.attr.check:
-    script_content += ctx.attr.check
-  script_content += "exit $result\n"
-  script_file = ctx.new_file(ctx.label.name+".bash")
-  ctx.file_action(output=script_file, executable=True, content=script_content)
-  # finalise the bazel options
-  ctx.file_action(output=bazelrc, content=_bazelrc)
+  else:
+    workspace_content += _basic_workspace.format()
+    workspace_content += register
+
+  workspace_file = ctx.actions.declare_file("WORKSPACE.in")
+  ctx.actions.write(workspace_file, workspace_content)
+  build_file = ctx.actions.declare_file("BUILD.in")
+  ctx.actions.write(build_file, ctx.attr.build)
+
+  targets = ["@" + ctx.workspace_name + "//" + ctx.label.package + t if t.startswith(":") else t for t in ctx.attr.targets]
+  output = "external/" + ctx.workspace_name + "/" + ctx.label.package
+  ctx.file_action(output=script_file, executable=True, content=_bazel_test_script_content.format(
```

Comment:
I think this would be more readable if the `format` call were extracted into a separate statement.